modem_cellular: add RSRP &RSRQ to U-blox SARA-R5 modem

emillindq opened the pull request:
This commit implements parsing of the CESQ extended signal quality AT command, extracting RSSI, RSRP and RSRQ, the latter two being relevant for LTE connections. It's used in the U-blox SARA-R5 modem instance. Furthermore, the IMEI and IMSI is extracted in the same modem.

A couple of comments:
- Not sure how to handle the "unknown" case, just returning without changing RSSI as done in `modem_cellular_chat_on_csq` feels wrong. Suggestions welcomed (maybe return -ENODATA if invalid?)
- Didn't add CESQ to periodic polling, which is done in the Quectel drivers. I don't see the point since you can't access RSSI/RSRP/RSRQ without calling the API, which runs them anyways

emillindq commented:
I'd like to request some help from @jeffwelder-ellenbytech or @bjarki-trackunit  regarding the failed pipeline. It fails because my added functions aren't used by the Quectel BG95 modem instance, which warning causes the sample "test" to fail.

It feels messy in the long run to ifdef all such parsing functions depending on which modem use them. I think modem_cellular.c is quite hard to navigate as it is due to the instantiation of the different modems, and now the API is added in the bottom where normally the device instantiation is. This feels like a larger discussion that should be taken elsewhere, because it feels like this structure of one giant .c file won't scale well when APIs are implemented, and more modems and AT commands are added. Do you agree?

I'm in the process of migrating from an out-of-tree mod of `ublox-sara-r4.c` to `modem_cellular.c`, and therefore would like to implement a lot of the missing APIs and things mentioned [here](https://github.com/zephyrproject-rtos/zephyr/issues/64405). However I feel that such discussion would be appropriate before it all blows up completely. Thanks!

bjarki-trackunit commented:
Hi, I know, I didn't look close enough to the actual API implementation that was added in modem_cellular.c and I'm looking into fixing it :)

bjarki-trackunit commented:
See https://github.com/zephyrproject-rtos/zephyr/pull/67066

emillindq commented:
> Hi, I know, I didn't look close enough to the actual API implementation that was added in modem_cellular.c and I'm looking into fixing it :) In short, API implementations will go into the struct device API pointers, and startup/periodic scripts will only be used if absolutely necessary, CESQ for example will to into its own script run synchronously when the API is called

@bjarki-trackunit The PR looks fine, I guess, but it doesn't address the problem in this PR (if that was the purpose). The thing is that `cesq_match` usage is not compiled in if compiling for Quectel model, and `csq_match` usage is not compiled in if compiling for Ublox modem. How to handle the warning?

Thanks!

bjarki-trackunit commented:
You should be able to rebase and add CESQ to the driver now that #67066 has been merged :)

emillindq committed:
modem_cellular: add RSRP & RSRQ to U-blox SARA-R5 modem

This commit implements parsing of the CESQ extended signal quality AT
command, extracting RSRP and RSRQ which is relevant for LTE connections.
It's used in the U-blox SARA-R5 modem instance. Furthermore, the IMEI,
IMSI is extracted in the same modem.

Signed-off-by: Emil Lindqvist <emil@lindq.gr>

emillindq commented:
@bjarki-trackunit please take a look!

bjarki-andreasen reviewed (dismissed):
Works as expected on the BG95, gets RSSI just fine, fails properly when trying to get RSRQ/RSRP (since it does not support AT+CESQ)

```
[00:00:22.941,131] <inf> modem_cellular: event script success
Signal RSSI -51
[00:00:23.980,895] <wrn> modem_chat: get_signal_cesq_chat_script: aborted
[00:00:23.980,957] <inf> modem_cellular: event script failed
Failed to get RSRP
[00:00:25.020,965] <wrn> modem_chat: get_signal_cesq_chat_script: aborted
[00:00:25.021,057] <inf> modem_cellular: event script failed
Failed to get RSRQ
[00:00:25.063,171] <inf> modem_cellular: event script success
Signal RSSI -51
```

fabiobaltieri commented on drivers/modem/modem_cellular.c:
```diff
@@ -1261,6 +1277,39 @@ static inline int modem_cellular_csq_parse_rssi(uint8_t rssi, int16_t *value)
+ */
+static inline int modem_cellular_cesq_parse_rsrp(uint8_t rsrp, int16_t *value)
+{
+	if (rsrp == 255) {
```
could you add few definitions for these? based on the description above I imagine they could be something like 
```c
#define CESQ_RSII_UNKNOWN 255
#define CESQ_RSRP_TO_DB(v) (-140 + v)
#define CESQ_RSRQ_TO_DB(v) (-20 + v / 2)
```

then the macro somehow self-describe the input and output unit together with the formula

emillindq replied:
Sure, however I don't think that necessarily makes it more clear. I think it's nice that the conversion functions are explicit about what's happening, so I don't have to go to a macro to find out. See what you think!

fabiobaltieri replied:
I think it's better, last word to the maintainers though, @rerickson1?
